---
title: "base_count"
output: html_document
date: "2023-04-22"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggplot2)
library(ggpubr)


```


MANDATORY DATA HANDLING
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#######################################################################################################
DATA HANDLING TO LOAD SEQUENCES, MAKE CONSENSUS, ETC.
#######################################################################################################
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Same across all analysis files



```{r}

################################################################################
#STEP 1: READING SEQUENCES IN FROM FASTA FORMAT INTO DATA FRAME FOR ANALYSIS#
################################################################################


# Read in the file
# readLines() creates a variable where each element contains a string for each line of the file


#PLEASE ENTER THE FILE PATH OF THE SEQUENCE FILE HERE
file_path <- "C:/Users/1044832/Desktop/Sequence_analysis/Reference_sequences/WT Wuhan Reference WIV04.fasta"
lines <- readLines(file_path)


#Initialize empty vectors for names and sequences
#Vectors need to be initialized prior to using in a loop
names_list <- c()
sequences_list <- c()

#Initialize new variables to hold the current names and sequence before adding them to the list
current_name <- ""
current_sequence <- ""

#names_list and current_name are initialised differently because current_name only needs to hold one name at a time (hence only needs to be a string) while all current_names will be added to names_list (hence it needs to be a character vector to hold multiple strings)


# Parse the lines
for (line in lines) { # goes through all the lines one by one 
  if (substr(line, 1, 1) == ">") { #substr separates out the part of the string that is specified, if first character of line is equal to > (a names line), then
    
    # New name, save previous sequence
    if (current_name != "") {
      names_list <- c(names_list, current_name)
      sequences_list <- c(sequences_list, current_sequence)
    }
    # Start new sequence with this name
    current_name <- gsub('>', '', line) # removing >
    current_sequence <- ""
  } else {
    # Add to current sequence
    current_sequence <- paste(current_sequence, line, sep="")
  }
}

# Save the last name and sequence
names_list <- c(names_list, current_name)
sequences_list <- c(sequences_list, current_sequence)

# Create a data frame with the name and sequence information
df <- data.frame(name = names_list, sequence = sequences_list)

# Cleaning up
# Removing unwanted variables
rm(list = c("current_name", "current_sequence", "line", "lines", "names_list", "sequences_list"))
# Garbage Collection = removing temporary memory
invisible(gc())


```


```{r}

################################################################################
#STEP 2: SEPARATING OUT EACH POSITION INTO ITS OWN ROW#
################################################################################

#identify the max length of the sequences
max_length <- max(str_length(df$sequence))

#Making all the sequences the same length for unnest_longer(), if sequences are shorter than the max length, add enough - at the end to make them equivalent
df$sequence <- str_pad(df$sequence, max_length, "right", "-")

#Separating each nucleotide
df$split <- strsplit(df$sequence, split="")

#Removing sequence column as its no longer needed
df.min <- subset(df, select = c("name", "split"))

#Making species names the column names
df.wide <- pivot_wider(df.min, names_from = name, values_from = split)

#Seperating the list into individual rows
data <- df.wide %>% unnest_longer(c(colnames(df.wide)))

#Cleaning environment
rm(list=setdiff(ls(), c("data", "df", "file_path")))
invisible(gc())


```


```{r}
################################################################################
#STEP 3: ADDING CODON POSITION INFORMATION#
################################################################################

#According to NCBI gene reference sequence for SARS-CoV-2, the open reading frame for SARS-CoV-2 WT starts at the second codon position.
wt_codon_annotations <- read.csv("WT_codon_annotation.csv")

#Adding position row into data data frame
data$position <- seq_along(1:nrow(data))

#Adding ORF and codon position information into separate columns.

#ORF1ab [266:21555] (-1 frameshift at 13468)
data$ORF1ab <- NA
data$ORF1ab[266:13468] <- rep(c(1, 2, 3))
data$ORF1ab[13468] <- "3, 1"
data$ORF1ab[13469:21555] <- rep(c(2, 3, 1))

#ORF1a [266:13468]
data$ORF1a <- NA
data$ORF1a[266:13468] <- rep(c(1, 2, 3))

#S [21563:25384]
data$S <- NA
data$S[21563:25384] <- rep(c(1, 2, 3))

#ORF3a [25393:26220]
data$ORF3a <- NA
data$ORF3a[25393:26220] <- rep(c(1, 2, 3))

#ORF3b [25765:26220]
data$ORF3b <- NA
data$ORF3b[25765:26220]<- rep(c(1, 2, 3))

#E [26245:26472]
data$E <- NA
data$E[26245:26472] <- rep(c(1, 2, 3))

#M [26523:27191]
data$M <- NA
data$M[26523:27191] <- rep(c(1, 2, 3))

#ORF6 [27202:27387]
data$ORF6 <- NA
data$ORF6[27202:27387] <- rep(c(1, 2, 3))

#ORF7a [27394:27759]
data$ORF7a <- NA
data$ORF7a[27394:27759] <- rep(c(1, 2, 3))

#ORF7b [27756:27887]
data$ORF7b <- NA
data$ORF7b[27756:27887] <- rep(c(1, 2, 3))

#ORF8 [27894:28259]
data$ORF8 <- NA
data$ORF8[27894:28259] <- rep(c(1, 2, 3))

#N [28274:29533]
data$N <- NA
data$N[28274:29533] <- rep(c(1, 2, 3))

#ORF9b [28284:28577]
data$ORF9b <- NA
data$ORF9b[28284:28577] <- rep(c(1, 2, 3))

#ORF9c [28734:28955]
data$ORF9c <- NA
data$ORF9c[28734:28955] <- rep(c(1, 2, 3))

#ORF10 [29558:29674]
data$ORF10 <- NA
data$ORF10[29558:29674] <- rep(c(1, 2, 3))


#Write into file
write.csv(data, file = "wt_codon_annotated.csv")




```

